peekingduck/pipeline/nodes/input/visual.py

```diff
@@ -0,0 +1,259 @@
+# limitations under the License.
+
+"""
+Reads inputs from multiple visual sources
+- image or video file on local storage
+- folder of images or videos
+- online cloud source
+- CCTV or webcam live feed
+"""
+
+from pathlib import Path
+from typing import Any, Dict, Optional, Union
+
+from peekingduck.pipeline.nodes.input.utils.preprocess import resize_image
+from peekingduck.pipeline.nodes.input.utils.read import VideoNoThread, VideoThread
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+
+class Node(AbstractNode):  # pylint: disable=too-many-instance-attributes
+    """Receives visual sources as inputs.
+
+    Inputs:
+        |none_input_data|
+
+    Outputs:
+        |img_data|
+
+        |filename_data|
+
+        |pipeline_end_data|
+
+        |saved_video_fps_data|
+
+    Configs:
+        filename (:obj:`str`): **default = "video.mp4"**. |br|
+            Filename of the MP4 file if media is exported. |br|
+            If source is a directory of files, then filename is the current file being
+            processed.
+        frames_log_freq (:obj:`int`): **default = 100**. |br|
+            Logs frequency of frames passed in CLI
+        mirror_image (:obj:`bool`): **default = False**. |br|
+            Flag to set extracted image frame as mirror image of input stream.
+        resize (:obj:`Dict`):
+            **default = { do_resizing: False, width: 1280, height: 720 }** |br|
+            Dimension of extracted image frame.
+        saved_video_fps (:obj:`int`): **default = 10**. |br|
+            FPS to be used to output the MP4 file after livestream is processed and
+            exported.  FPS is dependent on running machine performance.
+        source (:obj:`int`):
+            **default = https://storage.googleapis.com/peekingduck/videos/wave.mp4**. |br|
+            Input source can be: |br|
+            - filename : local image or video file |br|
+            - directory name : all media files will be processed |br|
+            - http URL for online cloud source : http[s]://... |br|
+            - rtsp URL for CCTV : rtsp://... |br|
+            - 0 for webcam live feed |br|
+            Refer to `OpenCV documentation
+            <https://docs.opencv.org/4.5.5/d8/dfe/classcv_1_1VideoCapture.html>`_
+            for more technical information.
+        threading (:obj:`bool`): **default = False**. |br|
+            Flag to enable threading when reading frames from camera.
+            The FPS can increase up to 30%.
+        buffer_frames (:obj:`bool`): **default = False**. |br|
+            Boolean to indicate if threaded class should buffer image frames.
+            If threading is True and output.media_writer is enabled, then
+            buffer_frames should be True to ensure output video is correctly
+            saved. One side effect of threading=True, buffer_frames=True is the
+            onscreen video display could appear laggy due to the buffering.
```

Comment:
Some thoughts about this:
1. Some configs such as `source` and `resize`+`mirror_image` are more likely to be used, while others like `buffer_frames` are more advanced and may intimidate new users. Especially since this is our only `input` node and everyone has to use this (unlike `dabble.statistics` for example). Could we group them somehow to make them more obvious, or re-arrange the order?
2. We have tech notes in the .yml file and I'm thinking it's time to move them out of the .yml into our docs. There are different scenarios where you should turn on certain configs, and I think it's important to list them here to avoid future confusion. See draft table below for a start:

![image](https://user-images.githubusercontent.com/33995028/158765730-c7506b5d-b884-49af-8ac8-8c5c0cdbea28.png)

3. If we are going with the above table, we should define what we mean by certain terms such as "delay" or "lag". Consider using GIFs to illustrate, if it is easier. See `draw.legend` node for example on how to include image in the docstring
4. There are a lot more combinations of `buffer_frames` and `threading` than illustrated above, would be good to raise errors when a combination that doesn't make sense is paired. Such as using `buffer_frames` with a recorded video input, or `threading` with a recorded video input (which gives an error from VideoThread at the moment)

Reply:
@liyier90 your thoughts are welcome too!

Reply:
1. Maybe can put a heading before the advanced configurations. Similar to the warning screen firefox does "Changing advanced configuration preferences can impact Firefox performance or security."
2. Does `live` apply to both streaming a mp4 file and a webcam feed? Might need to disclaimer about internet/streaming speed.
3. Good to be clear about the description of the observed side effect. Lag or laggy may be ambiguous sometimes
4. May need to consider the effort required, since validity of the config depends on input type. Example: it might be a folder of images and videos or an URL of an image

Reply:
Rewrite `threading` and `buffer_frames` description to recommend when to enable them.
Regroup `input.visual` configs, but unfortunately current RTD template does not allow for nice group (see image below).

![Screen Shot 2022-03-19 at 10 35 53 PM](https://user-images.githubusercontent.com/13075447/159125434-135ae531-b83c-434f-a1aa-e055a27bd836.png)

Reply:
Recommended `input.visual` config: the green box.

![input_visual_config](https://user-images.githubusercontent.com/13075447/159127757-6db4b623-f4e9-427c-b92a-b4c061177773.png)

p.s. Video file : True, True will not have laggy video 'coz each frame is buffered once, no duplicate unlike webcam.

Q: where to put this diagram... FAQ?

Reply:
1. **Placement of diagram** - Prefer putting the definitions and links to "laggy" etc, and this diagram within this node's docstrings, just like in `dabble.statistics`, since this shows how users should be using the advanced configs. I don't think it's required in the FAQ for the time being, unless we start having more questions from our users.  Possibly, place under the advanced configs section and the config explanations for `buffer_frames` and `threading` can be shortened and refer to this diagram.

2. **Contents of diagram** - Can consider using this: https://www.tablesgenerator.com/text_tables to generate the rst table. `-` may not be the best symbol to represent "normal behavior", at first glance it looks like it's not valid.

3. There are 3 scenarios (please verify) for determining the `saved_video_fps` config : (a) Video file/image -> Ignored, it uses FPS of original video/N.A. for images; (b) Threading (True) + Buffering (True) -> FPS of the CCTV/webcam; (c) Threading (True) + Buffering (False) -> Estimate the FPS based on machine performance, suggest doing a run for a few minutes to get the average FPS, and use that in this config. Either can list these options within this config (point form/another table) or merge it with the main diagram.

4. Why do we call the green box the recommended behaviour? I believe we're still keeping both options as `False` by default?

Reply:
> Placement of diagram
1. Will add diagram into the docs. For "laggy", best is to use the `wave_vs_wave.mp4` to show a normal vs lagged video. We can host the file (~4 MB) in GCP and link to it.

> Contents of diagram 
2. Agree. Have updated the diagram (see below): use tick to mean "normal behavior", `+` to indicate increased FPS. If want to replace diagram with a text-based table, then will refactor all the symbols to something showable on text.

> There are 3 scenarios (please verify) for determining the saved_video_fps config
3. The original text taken off the old `input.live` was mean to address user configuration of this parameter. But internally, this is overridden by the video FPS if available, and this was formerly undocumented. Will extend the description to elaborate all the internal workings of how PKD use/override this parameter.

> Why do we call the green box the recommended behaviour? I believe we're still keeping both options as False by default?
My bad, should have clarified: the normal use-case is for user to ignore Threading/Buffering. But if user really wants to play with Threading/Buffering, then the green box is the only sensible choice, since the others are either as good as normal behavior or worse (red boxes).

![input_visual_config](https://user-images.githubusercontent.com/13075447/159210761-41730253-cb3a-42b1-8d8e-0a5218f9f3af.png)